File: Taggers/java/src/Tagger.java
import com.sun.net.httpserver.HttpServer;

import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Created by anubhav on 11/17/15.
 */
public class Tagger {
    Status status;
    long ctr = 0;
    Boolean do_started_ack(HashMap<String, Object> args)
    {
        status = Status.STARTED;
        return true;
    }

    Boolean do_training_ack(HashMap<String, Object> args)
    {
        status = Status.TRAINING;
        return true;
    }

    Boolean do_active_ack(HashMap<String, Object> args)
    {
        status = Status.ACTIVE;
        return true;
    }

    Boolean do_disabled_ack(HashMap<String, Object> args)
    {
        status = Status.DISABLED;
        return true;
    }

    Object compute(HashMap<String, Object> args)
    {
        if (status != Status.ACTIVE) {
            return (Boolean)false;
        }
        return (Boolean)false;
    }

    Object train(HashMap<String, Object> args)
    {
        return true;
    }

    protected ArrayList<String> paths;
    public Object cmds(String cmd, HashMap<String, Object> args)
    {
        if (cmd.toLowerCase().equals("started_ack")) {
            return do_started_ack(args);
        } else if (cmd.toLowerCase().equals("training_ack")) {
            return do_training_ack(args);
        } else if (cmd.toLowerCase().equals("active_ack")) {
            return do_active_ack(args);
        } else if (cmd.toLowerCase().equals("disabled_ack")) {
            return do_disabled_ack(args);
        } else {
            return "Unknown";
        }
    }

    public String handle(String cmd, HashMap<String, Object> args)
    {
        return "";
    }
    public Tagger(int port) {
        this.port = port;
        paths = new ArrayList<String>() {
            {
                add("started_ack");
                add("training_ack");
                add("active_ack");
                add("disabled_ack");
            }
        };
    }

    public enum Status
    {
        DISABLED,
        STARTED,
        TRAINING,
        ACTIVE
    }
    int port = 8000;
    public void run() throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress(port), 0);
        for (String path: paths) {
            System.out.println("Adding path: " + path);
            server.createContext("/"+path, new RequestHandler(path, this));
        }
        server.setExecutor(null);
        server.start();
    }

    public long tick()
    {
        long c = ctr;
        ++ctr;
        return c;
    }

}

File: Taggers/python/Tagger.py
#Python base class for building a tagger

from BaseHTTPServer import BaseHTTPRequestHandler
from BaseHTTPServer import HTTPServer
import urlparse
DISABLED, STARTED, TRAINING, ACTIVE = range(4)
class Tagger(object):
    handler = None
    status = DISABLED
    
    def do_started_ack(args):
        status = STARTED
        return True
    
    def do_training_ack(args):
        status = TRAINING
        return True
    
    def do_active_ack(args):
        status = ACTIVE
        return True
    
    def do_disabled_ack(args):
        status = DISABLED
        return True

    def train(self, args=None):
        raise NotImplementedError

    # Tag based on provided features, return class/value
    def compute(self, args=None):
        raise NotImplementedError

    cmds = {
        "started_ack": do_started_ack,
        "training_ack": do_training_ack,
        "active_ack": do_active_ack,
        "disabled_ack": do_disabled_ack
    }
    def __init__(self, args=None):
        try:
            self.handler = args['handler']
        except KeyError:
            raise KeyError("args['handler'] is a required argument")
        pass
    def run(self, args=None):
        server = HTTPServer(('localhost', 8080), self.handler)
        print 'Starting server, use <Ctrl-C> to stop'
        server.serve_forever()

'''
    class GetHandler(BaseHTTPRequestHandler):
        
        def do_GET(self):
            parsed_path = urlparse.urlparse(self.path)
            res = ''
            # extract data, perform computation and then return result
            # Also handle start/stop state requests
            
            self.send_response(200)
            self.end_headers()
            self.wfile.write(res)
            return
'''